File: python/sorting/quick_sort.py
import random

def partition(input_sequence, Left, Right):

            limit = input_sequence[Left]

            Less = Left + 1

            More = Less

            while More != Right:

                if input_sequence[ More ] < limit:

                    input_sequence[ Less ], input_sequence[ More ] = input_sequence[ More ], input_sequence[ Less ]

                    Less += 1

                More += 1

            pivot = Less - 1

            input_sequence[ Left ], input_sequence[ pivot ] = input_sequence[ pivot ], input_sequence[ Left ]

            return pivot

def partition3(input_sequence,Left, Right):

            part = partition( input_sequence, Left, Right )

            m1 = part

            m2 = part

            i = Right - 1

            while m2 < i:

                if input_sequence[ i ] == input_sequence[ part ]:

                    input_sequence[ m2 ], input_sequence[ i ] = input_sequence[ i ], input_sequence[ m2 ]

                    m2 += 1

                else:

                    i -= 1

            return m1, m2

       

def randomized_quick_sort(input_sequence, Left, Right ):

            if Right - Left < 2:

                return

            part = random.randint( Left, Right-1 )

            input_sequence[ Left ], input_sequence[ part ] = input_sequence[ part ], input_sequence[ Left ]

            m1, m2 = partition3( input_sequence, Left, Right )

            randomized_quick_sort( input_sequence, Left, m1 )

            randomized_quick_sort( input_sequence, m2, Right )

if __name__ == '__main__':

        no_of_inputs = input('Enter no. of inputs')

        input_sequence= list( map( int, input('Enter numbers to be sort separated by one spacing').split() ))

        Left = 0

        Right = len(input_sequence)    

        randomized_quick_sort( input_sequence , Left , Right )

        for x in input_sequence:

            print( x, end=" " )

File: go/sorting/quicksort.go
package quicksort

import (
	"fmt"
	"math/rand"
)

//QuickSort uses quick sort to sort an array of integers
func QuickSort(list []int) []int {
	if len(list) < 2 {
        return list
    }
     
    left, right := 0, len(list)-1
     
    pivot := rand.Int() % len(list)
     
    list[pivot], list[right] = list[right], list[pivot]
     
    for i, _ := range list {
        if list[i] < list[right] {
            list[left], list[i] = list[i], list[left]
            left++
        }
    }
     
    list[left], list[right] = list[right], list[left]
     
    QuickSort(list[:left])
    QuickSort(list[left+1:])
     
    return list
}

func main() {
	list := []int{4324, 442, 422, 6931, 356, 84722, 1, 3452, 24, 13, 89, 0, -111}

	sortedArray := QuickSort(list)

	fmt.Println(sortedArray)
}
